README.md
# py3bt_tracker

py3bt_tracker is a cross-platform standalone ephemeral tracker for torrents written in Python3.

The development emphasis is on zero-configuration "just works" software.

Currently implemented using a multidimentional array in memory and the Tornado Web Server (http).

google_define('ephemeral') #returns: adjective - lasting for a very short time, transient, short-lived, brief

## Key Features:

- Works
- Prebuilt binaries
- No external dependencies or configuration required
- Permissive license
- Standards compliant (including IPv6 support)
- Easy to rebuild from source if so inclined (see release_notes.txt)

## Download:

```
Latest Release: 1.0.0
Development: stopped. Open an "issue" or pm me for a requested use case/feature or bug
```
Click [here](https://github.com/gdiaz384/py3bt_tracker/releases) or on "releases" at the top to download the latest version.

## Typical Usage Guide:

1. double click on the .exe for your os
2. point your torrents to http://192.168.1.50:6969/announce where 192.168.1.50 gets substituted for your IP obtained from ipconfig (for lans)
3. Note: If the firewall annoyance pops up, add as an exception.

## Release Notes Summary:

- By default, runs on port 6969 with a check-in interval of 4 minutes and database lifespan of 1 hour.
- Currently no scrape/statistics/private/logging/obfuscation features
- Intended use case is for LANs. Will also work on the internet, but not currently optimized for it.
- If there's any interest in doing so, I'll do a version 2.0.0 for optimized use on the internetwork system, with UDP/obfuscation support, since that's the natural extension to this project. Message me over git.

## Advanced Usage Guide:
```
Syntax: py3bt_tracker {--port=6969} {--client_request_interval=4} {--enable_obfuscation}
Syntax: py3bt_tracker {-p=9000} {-i=30} {-o}
>py3bt_tracker --help
>py3bt_tracker
>py3bt_tracker --port=6969
>py3bt_tracker -p=6969
>py3bt_tracker --port=9000 --client_request_interval=4
>py3bt_tracker -p=80 -i=15
Windows (binary): >start "" C:\Users\User\Downloads\py3bt_tracker.exe --port=9000
Windows (pyscript): >start "" python C:\Users\User\Downloads\py3bt_tracker.py -i=30
```

Notes: obfuscation not currently implemented, planned for 2.0.0 if there's enough interest

## License:
Pick your License: any version of GPL/BSD/MIT or Apache

If I get any questions on licensing, I'm changing this to "beerware" and will refuse to elaborate further.


release_notes.txt
Release Notes:
-Requires Python and Tornado web server (for python), see How To Build section for more details
-Runs on port 6969, with a requested check-in interval of 4 minutes by default (change via command line)
-Intended use case is in a LAN enviornment, will also work on the internet, but not currently optimized for it.
-currently no scrape/statistics/private/logging features 
 UDP and obfuscation might be added at some point

Notice: Although it should work, py3bt_tracker is not currently optimized for performance/security in open/untrusted WAN enviornments (the internet) and extremely heavy loads. Should work well otherwise. For linux enviornments, consider building OpenTracker from source over at http://erdgeist.org/arts/software/opentracker

Read the "Stuff Not To Spec" section and "Todo List" to determine if this is suitable for your enviornment.

Usage:
double click on it
Detailed Usage: see advanced usage in the readme.md or run py3bt_tracker.exe --help
To close/stop it: ctrl+c or click on the little x at the top right of the command prompt window
To log output: py3bt_tracker.exe > %userprofile%\desktop\py3bt_tracker.log.txt (very bare atm)

Motivation/goal:
-opentracker is posix only due to fat library dependency :'(  
-same with and all other trackers I've found, either non-functional or dependencies everywhere or both why x.x
-This one is a cross platform non-equivalent with an emphasis on reduced performance and fewer (end-user) dependencies, 
as in: compiled already that doesn't need linux/webserver/php/sql/vsredists/boost/ides/vstudio/knowinghowtocompilesoftware/webengine etc

Spec available at:
https://wiki.theory.org/BitTorrentSpecification 
https://wiki.theory.org/BitTorrent_Tracker_Protocol
http://bittorrent.org ->developers

Stuff Not To Spec:
-first announce to tracker must have "started" event item. Why not enforced? this makes sense in order to keep server-side statistics, ephemeral ones don't however so enforcing this is pointless
-client numwant not respect respected, set to a max of 50 instead. Why not respected? defaulting to 50 is the spec and even 30 is plenty according to implementation notes in spec, crazy clients that want 200 are crazy
-scrape option not implemented. Why not? doesn't make sense for ephemeral trackers
-option to enable registered info-hashes only. Why not? cuz ephemeral and also complicated to implement
-option to only return peer list to 'registered' peers. Why not? cuz idea is stupid and also very complicated to implement
-beep31 why not? that's more for web servers in general to respond to clients, non really for actual trackers
-trackerID. Why not? i'm assuming this is like an xsrf token to identify the peer. very limited use cases and coding depends strongly on how the tracker is intended to be used
-beep6 why not? not sure if it's widely supported enough to bother implementing, and also looks complicated, and not sure it even makes sense since anyone sniffing the line could always just do a replay attack and receive the peer list right? And anyway, TLS is a better solution to passive sniffing. Mere obfuscation would help only if data is stored in a db for anaysis later, but still any 'decrypted' obfuscation would be very damning evidence so... If this is important, then it makes more sense to seed the db with fake peers to contaminate the evidence preemtively. I could totally code this if bored enough one day.

Questions about why this is so poorly written:
-my first python program
-my first useful program, as in program that was actually useful at solving a novel problem not solved before (standalone tracker for windows)
-emphasis is on working/legibility, not running super fast (not optimized for use on internetz)
-not sure exactly how to reference python class variables
-not sure how to create dynamic variables in py3, and py2 method using exec() was a hack
-and not sure how to print/message_pass/append binary, as in native binary, cmd doesn't seem to like it so can't rencode back to valid data after using str(binary_data)

Version 1 Todo list:
-incomplete and complete should increment/decrement -currently buggy (cosmetic tho)
-(bug) ipv6 integer ports are messed up ii35104ee
-maybe implement logging, maybe 

-Version 2 Todo list:
-add required response delay before giving out peers or adding to main peer db to raise the bar on potential for abuse in untrustworthy enviornments
-improve performance for use in WANs (production quality code optimizations + fully async web server)
-also add obfuscation feature (using the ipaddress library), and make wipe old peers/db optional (essentially have a reliably mode and a screwy mode filled with wrong info)
-maybe v2 should have a "high integrity mode" and a deliberate "low integrity mode"
-add udp tracker support (beep 15) -Does Tornado even know what UDP is? maybeh, or just import udp support from another project

NOT on todo list:
scrape functionality -cuz ephemeral
private tracker feature - registered infohashes -cuz ephemeral
private tracker feature - registered peers -cuz stupid


How To Build:

Stop reading this and go get binaries available under "releases".

https://github.com/gdiaz384/py3bt_tracker/releases

Seriously.


##don't read past this line##


If you're still reading...let's get started.

Windows Vista,7,8,81,10 64-bit (x64)

google -> python -> download -> different OS -> Windows -> Python 3.4.3 - 2015-02-25 -> Windows x86-64 MSI installer
https://www.python.org/ftp/python/3.4.3/python-3.4.3.amd64.msi
(required) add to path 
(optional) compile py to binary files 
http://tdm-gcc.tdragon.net/
http://sourceforge.net/projects/tdm-gcc/files/TDM-GCC%20Installer/tdm64-gcc-5.1.0-2.exe/download
log out and log back in or reboot computer
run->cmd
>pip install tornado
>pip install pyinstaller
google -> py3bt_tracker -> github -> click on py3bt_tracker.py -> raw -> ctrl+S
or
https://raw.githubusercontent.com/gdiaz384/py3bt_tracker/master/py3bt_tracker.py 
pyinstaller --version

debug/torn_write.py
import hashlib
import urllib.parse
import binascii
import time

import os
#import random

#import os.path
#import sys
#import uuid
#import logging
#import math
#import optparse
#from optparse import OptionParser
#import re
#import base64

#from py3bencode import bencode

#import tornado
import tornado.ioloop
import tornado.web
import tornado.httpserver
import tornado.httputil
#import tornado.escape
#import tornado.gen
#from tornado.concurrent import Future
from tornado.options import define, options, parse_command_line

class MainHandler(tornado.web.RequestHandler):
    def get(self):
        peerlist_hex_string='c0a8006a84f3c0a80064e9c7c0a800960c80c0a8006c33f1c0a8006a1b4b'
        peerlist_binary=binascii.a2b_hex(peerlist_hex_string)
        response=peerlist_binary
        print(response)
        
        #response='peers'
        #self.set_header('Content-Type', 'application/octet-stream')
        self.write('string')
        self.write(response)
        self.flush()
        #output for dictionary type:
        #dictionary peer list (multiple strings + bitstreams, difficult to pass back-forth through messages)
        #dictionary w/o peer list 1 string
        #output for compact type:
        #only 1 type but needs: 1string+1bitstream (peers) (can append e at the end)
        #could return...each component, and a peerlist, then write it dynamically as chunks in main response
        #return components + peer string, (bitstream list)
    
def main():
    parse_command_line()
    global tracker_db
    #tracker_db = Database()
    app = tornado.web.Application([
            (r"/announce.*", MainHandler),
            ])
    app.listen(8080)
    tornado.ioloop.IOLoop.current().start()

if __name__ == "__main__":
    main()

debug/fixWindowsCMD.py
import sys
if sys.platform == "win32":
    import codecs
    from ctypes import WINFUNCTYPE, windll, POINTER, byref, c_int
    from ctypes.wintypes import BOOL, HANDLE, DWORD, LPWSTR, LPCWSTR, LPVOID

    original_stderr = sys.stderr

    # If any exception occurs in this code, we'll probably try to print it on stderr,
    # which makes for frustrating debugging if stderr is directed to our wrapper.
    # So be paranoid about catching errors and reporting them to original_stderr,
    # so that we can at least see them.
    def _complain(message):
        print >>original_stderr, message if isinstance(message, str) else repr(message)

    # Work around <http://bugs.python.org/issue6058>.
    codecs.register(lambda name: codecs.lookup('utf-8') if name == 'cp65001' else None)

    # Make Unicode console output work independently of the current code page.
    # This also fixes <http://bugs.python.org/issue1602>.
    # Credit to Michael Kaplan <http://www.siao2.com/2010/04/07/9989346.aspx>
    # and TZOmegaTZIOY
    # <http://stackoverflow.com/questions/878972/windows-cmd-encoding-change-causes-python-crash/1432462#1432462>.
    try:
        # <http://msdn.microsoft.com/en-us/library/ms683231(VS.85).aspx>
        # HANDLE WINAPI GetStdHandle(DWORD nStdHandle);
        # returns INVALID_HANDLE_VALUE, NULL, or a valid handle
        #
        # <http://msdn.microsoft.com/en-us/library/aa364960(VS.85).aspx>
        # DWORD WINAPI GetFileType(DWORD hFile);
        #
        # <http://msdn.microsoft.com/en-us/library/ms683167(VS.85).aspx>
        # BOOL WINAPI GetConsoleMode(HANDLE hConsole, LPDWORD lpMode);

        GetStdHandle = WINFUNCTYPE(HANDLE, DWORD)(("GetStdHandle", windll.kernel32))
        STD_OUTPUT_HANDLE = DWORD(-11)
        STD_ERROR_HANDLE = DWORD(-12)
        GetFileType = WINFUNCTYPE(DWORD, DWORD)(("GetFileType", windll.kernel32))
        FILE_TYPE_CHAR = 0x0002
        FILE_TYPE_REMOTE = 0x8000
        GetConsoleMode = WINFUNCTYPE(BOOL, HANDLE, POINTER(DWORD))(("GetConsoleMode", windll.kernel32))
        INVALID_HANDLE_VALUE = DWORD(-1).value

        def not_a_console(handle):
            if handle == INVALID_HANDLE_VALUE or handle is None:
                return True
            return ((GetFileType(handle) & ~FILE_TYPE_REMOTE) != FILE_TYPE_CHAR
                    or GetConsoleMode(handle, byref(DWORD())) == 0)

        old_stdout_fileno = None
        old_stderr_fileno = None
        if hasattr(sys.stdout, 'fileno'):
            old_stdout_fileno = sys.stdout.fileno()
        if hasattr(sys.stderr, 'fileno'):
            old_stderr_fileno = sys.stderr.fileno()

        STDOUT_FILENO = 1
        STDERR_FILENO = 2
        real_stdout = (old_stdout_fileno == STDOUT_FILENO)
        real_stderr = (old_stderr_fileno == STDERR_FILENO)

        if real_stdout:
            hStdout = GetStdHandle(STD_OUTPUT_HANDLE)
            if not_a_console(hStdout):
                real_stdout = False

        if real_stderr:
            hStderr = GetStdHandle(STD_ERROR_HANDLE)
            if not_a_console(hStderr):
                real_stderr = False

        if real_stdout or real_stderr:
            # BOOL WINAPI WriteConsoleW(HANDLE hOutput, LPWSTR lpBuffer, DWORD nChars,
            #                           LPDWORD lpCharsWritten, LPVOID lpReserved);

            WriteConsoleW = WINFUNCTYPE(BOOL, HANDLE, LPWSTR, DWORD, POINTER(DWORD), LPVOID)(("WriteConsoleW", windll.kernel32))

            class UnicodeOutput:
                def __init__(self, hConsole, stream, fileno, name):
                    self._hConsole = hConsole
                    self._stream = stream
                    self._fileno = fileno
                    self.closed = False
                    self.softspace = False
                    self.mode = 'w'
                    self.encoding = 'utf-8'
                    self.name = name
                    self.flush()

                def isatty(self):
                    return False

                def close(self):
                    # don't really close the handle, that would only cause problems
                    self.closed = True

                def fileno(self):
                    return self._fileno

                def flush(self):
                    if self._hConsole is None:
                        try:
                            self._stream.flush()
                        except Exception as e:
                            _complain("%s.flush: %r from %r" % (self.name, e, self._stream))
                            raise

                def write(self, text):
                    try:
                        if self._hConsole is None:
                            if isinstance(text, unicode):
                                text = text.encode('utf-8')
                            self._stream.write(text)
                        else:
                            if not isinstance(text, unicode):
                                text = str(text).decode('utf-8')
                            remaining = len(text)
                            while remaining:
                                n = DWORD(0)
                                # There is a shorter-than-documented limitation on the
                                # length of the string passed to WriteConsoleW (see
                                # <http://tahoe-lafs.org/trac/tahoe-lafs/ticket/1232>.
                                retval = WriteConsoleW(self._hConsole, text, min(remaining, 10000), byref(n), None)
                                if retval == 0 or n.value == 0:
                                    raise IOError("WriteConsoleW returned %r, n.value = %r" % (retval, n.value))
                                remaining -= n.value
                                if not remaining:
                                    break
                                text = text[n.value:]
                    except Exception as e:
                        _complain("%s.write: %r" % (self.name, e))
                        raise

                def writelines(self, lines):
                    try:
                        for line in lines:
                            self.write(line)
                    except Exception as e:
                        _complain("%s.writelines: %r" % (self.name, e))
                        raise

            if real_stdout:
                sys.stdout = UnicodeOutput(hStdout, None, STDOUT_FILENO, '<Unicode console stdout>')
            else:
                sys.stdout = UnicodeOutput(None, sys.stdout, old_stdout_fileno, '<Unicode redirected stdout>')

            if real_stderr:
                sys.stderr = UnicodeOutput(hStderr, None, STDERR_FILENO, '<Unicode console stderr>')
            else:
                sys.stderr = UnicodeOutput(None, sys.stderr, old_stderr_fileno, '<Unicode redirected stderr>')
    except Exception as e:
        _complain("exception %r while fixing up sys.stdout and sys.stderr" % (e,))


    # While we're at it, let's unmangle the command-line arguments:

    # This works around <http://bugs.python.org/issue2128>.
    GetCommandLineW = WINFUNCTYPE(LPWSTR)(("GetCommandLineW", windll.kernel32))
    CommandLineToArgvW = WINFUNCTYPE(POINTER(LPWSTR), LPCWSTR, POINTER(c_int))(("CommandLineToArgvW", windll.shell32))

    argc = c_int(0)
    argv_unicode = CommandLineToArgvW(GetCommandLineW(), byref(argc))

    argv = [argv_unicode[i].encode('utf-8') for i in xrange(0, argc.value)]

    if not hasattr(sys, 'frozen'):
        # If this is an executable produced by py2exe or bbfreeze, then it will
        # have been invoked directly. Otherwise, unicode_argv[0] is the Python
        # interpreter, so skip that.
        argv = argv[1:]

        # Also skip option arguments to the Python interpreter.
        while len(argv) > 0:
            arg = argv[0]
            if not arg.startswith(u"-") or arg == u"-":
                break
            argv = argv[1:]
            if arg == u'-m':
                # sys.argv[0] should really be the absolute path of the module source,
                # but never mind
                break
            if arg == u'-c':
                argv[0] = u'-c'
                break

    # if you like:
    sys.argv = argv

debug/parseQuery.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

import codecs
import binascii
import urllib.parse
import hashlib
import base64
import sys
import random

#import codecs
#codecs.register(lambda name: codecs.lookup('utf-8') if name == 'cp65001' else None)
#codecs.register(lambda name: codecs.lookup('utf-8') if name == 'cp437' else None)
#import fixWindowsCMD.py

random.seed()

#raw query
query="info_hash=%124Vx%9A%BC%DE%F1%23Eg%89%AB%CD%EF%124Vx%9A&peer_id=-UT2210-%b9b%2cs%c3%15%c6Q%3cu%95%96&port=18584&uploaded=0&downloaded=0&left=109603551&corrupt=0&key=0E09C995&event=started&numwant=200&compact=1&no_peer_id=1"
print("query: " + query)

#print the number of items in uri
queryItems=len(query.split("&"))
print("queryItems: " + str(queryItems))

#place each item into a dictionary container
client_rawRequestContainer=dict({})
for i in range(queryItems):
    print (str(i)+": "+(query.split("&")[i]))
    temp=query.split("&")[i]
    client_rawRequestContainer.update([temp.split("=")])

#initalize temp variables
invalid="invalid"
client_info_hash=invalid
client_peer_id=invalid
client_port=invalid
client_uploaded=invalid
client_downloaded=invalid
client_left=invalid
client_compact=invalid
client_no_peer_id=invalid
client_event=invalid
client_ip=invalid
client_numwant=invalid
client_key=invalid
client_trackerid=invalid
client_corrupt=invalid

#search dictionary to make sure the item is present, then place into temporary variables, error out on required variables
if ('info_hash' in client_rawRequestContainer) == (True):
    client_info_hash=client_rawRequestContainer["info_hash"]
else:
    print("error: info_hash not present")

if ('peer_id' in client_rawRequestContainer) == (True):
    client_peer_id=client_rawRequestContainer["peer_id"]
else:
    print("error: peer_id not present")
	
if ('port' in client_rawRequestContainer) == (True):
    client_port=client_rawRequestContainer["port"]
else:
    print("error: port not present")
	
if ('uploaded' in client_rawRequestContainer) == (True):
    client_uploaded=client_rawRequestContainer["uploaded"]
else:
    print("error: uploaded not present")
	
if ('downloaded' in client_rawRequestContainer) == (True):
    client_downloaded=client_rawRequestContainer["downloaded"]
else:
    print("error: downloaded not present")
	
if ('left' in client_rawRequestContainer) == (True):
    client_left=client_rawRequestContainer["left"]
else:
    print("error: left not present")
	
if ('compact' in client_rawRequestContainer) == (True):
    client_compact=client_rawRequestContainer["compact"]
else:
    print("error: compact not present")
	
if ('no_peer_id' in client_rawRequestContainer) == (True):
    client_no_peer_id=client_rawRequestContainer["no_peer_id"]
else:
    print("error: no_peer_id not present")
	
if ('event' in client_rawRequestContainer) == (True):
    client_event=client_rawRequestContainer["event"]
else:
    print("error: event not present")

if ('ip' in client_rawRequestContainer) == (True):
    client_ip=client_rawRequestContainer["ip"]
else:
    print("error: ip not present")
	
if ('numwant' in client_rawRequestContainer) == (True):
    client_numwant=client_rawRequestContainer["numwant"]
else:
    print("error: numwant not present")
	
if ('key' in client_rawRequestContainer) == (True):
    client_key=client_rawRequestContainer["key"]
else:
    print("error: key not present")
	
if ('trackerid' in client_rawRequestContainer) == (True):
    client_trackerid=client_rawRequestContainer["trackerid"]
else:
    print("error: trackerid not present")
	
if ('corrupt' in client_rawRequestContainer) == (True):
    client_corrupt=client_rawRequestContainer["corrupt"]
else:
    print("error: corrupt not present")

#print_client_info
print("client_info_hash: "+client_info_hash)
print("client_peer_id: "+client_peer_id)
print("client_port: "+client_port)
print("client_uploaded: "+client_uploaded)
print("client_downloaded: "+client_downloaded)
print("client_left: "+client_left)
print("client_compact: "+client_compact)
print("client_no_peer_id: "+client_no_peer_id)
print("client_event: "+client_event)
print("client_ip: "+client_ip)
print("client_numwant: "+client_numwant)
print("client_key: "+client_key)
print("client_trackerid: "+client_trackerid)
print("client_corrupt: "+client_corrupt)


print()
raw_info_hash=b'\x12\x34\x56\x78\x9a\xbc\xde\xf1\x23\x45\x67\x89\xab\xcd\xef\x12\x34\x56\x78\x9a'

#so the info hash can't display in raw form, so encode it to a utf-8 bytestream (displayable)
#and then convert it to a string so it can be appended to the other string "raw_info_hash: "
#still won't be displayed exactly right but close enough, and it's stored correctly internally
print("raw_info_hash: "+str(raw_info_hash))
print("decoded_raw_info_hash"+(str(binascii.b2a_hex(raw_info_hash))))
print("length: "+str(len(binascii.b2a_hex(raw_info_hash))))
#40 hex digits means 20 bytes long

print()
encoded_info_hash=urllib.parse.quote_from_bytes(raw_info_hash,safe='')
print("url encoded_info_hash: " +encoded_info_hash)
#print("encoded_info_hash: " +encoded_info_hash.encode('utf-8'))

print()
decoded_info_hash=binascii.b2a_hex(urllib.parse.unquote_to_bytes(encoded_info_hash))
print("decoded_info_hash: "+str(decoded_info_hash))

#print()
#encoded_info_hash is the URL % encoded format
#now try for the non % format, failed x.x
#info_hash_data=base64.urlsafe_b64decode(encoded_info_hash)
#print("info_hash_data: "+str(info_hash_data))

#print("decoded_info_hash: "+str(decoded_info_hash.encode('utf-8')))
#print("raw_info_hash: "+str(raw_info_hash))
#encoded_info_hash=urllib.parse.quote(raw_info_hash,safe='')
#print("encoded_info_hash: "+str(encoded_info_hash))
#encoded_info_hash=urllib.parse.unquote(raw_info_hash, safe='')
#print("unencoded_info_hash: "+str(encoded_info_hash))

print()
raw_unencoded_peer_id='-UT2210-b9b2csc315c6Q3cu9596'
print("raw_unencoded_peer_id: "+str(raw_unencoded_peer_id))

prefix=raw_unencoded_peer_id[:raw_unencoded_peer_id.rfind('-')+1]
raw_unencoded_peer_id_part=raw_unencoded_peer_id[raw_unencoded_peer_id.rfind('-')+1:]
print(str(len(urllib.parse.unquote_to_bytes(raw_unencoded_peer_id_part))))
url_encoded_raw_peer_id_part=urllib.parse.quote(raw_unencoded_peer_id_part)
print("url_encoded_raw_peer_id_part: "+str(url_encoded_raw_peer_id_part))
#decoded_raw_peer_id=binascii.b2a_hex(urllib.parse.unquote_to_bytes(url_encoded_raw_peer_id))
decoded_raw_peer_id=prefix+str(url_encoded_raw_peer_id_part)
print("decoded_raw_peer_id: "+str(len(decoded_raw_peer_id))+": "+str(decoded_raw_peer_id))

#standard peer id
#aria's random hex method (also use by bram and others)
#shadow's method


#decode into bytes, make sure prefix+bytes = 20
#if prefix not present, then just convert to bytes and make sure =20
#Then stored form is no longer url encoded and ready for transmission
#make sure when creating hash that it's from the binary form
#do this when storing info_hash, peer_id and (per peer) keys

print()
print()
raw_encoded_url_string='-UT2210-%b9b%2cs%c3%15%c6Q%3cu%95%96'

print("raw_encoded_url_string: "+str(raw_encoded_url_string))
#check if string has seperate parts, if so decode only the second part
#a length of 24 hex digits means 12 bytes, 12 bytes of hex + prefix = 20 bytes if prefix is 8 bytes long
#35-3 (from hex to string conversion) is 32-8 =24, 24/2 = 12+8 = 20 bytes
#Note: the version could be encoded into the binary form itself instead of shown raw in the url (Aria does this)
if raw_encoded_url_string.rfind('-') == -1:
	decoded_url_string=urllib.parse.unquote_to_bytes(raw_encoded_url_string)
	#print(str(decoded_url_string))
	if len(decoded_url_string) != 20:
		print('error')
	temp=str(binascii.b2a_hex(decoded_url_string))
	#print(temp)
	decoded_url_string=temp[2:len(temp)-1]
elif raw_encoded_url_string.rfind('-') != -1:
	prefix=raw_encoded_url_string[:raw_encoded_url_string.rfind('-')+1]
	#print(str(len(prefix))+prefix)
	peer_id_part=raw_encoded_url_string[raw_encoded_url_string.rfind('-')+1:]
	decoded_url_string=urllib.parse.unquote_to_bytes(peer_id_part)
	#print(str(decoded_url_string))
	if len(prefix)+len(decoded_url_string) != 20:
		print('error')
	temp=str(binascii.b2a_hex(decoded_url_string))
	temp=temp[2:len(temp)-1]
	#print(temp)
	decoded_url_string=prefix+temp
print("decoded_url_string: "+decoded_url_string)


print()
raw_encoded_url_string='A2%2D1%2D19%2D0%2D%3DQ%C8%D3%7DS%BD%DA%95%25'

print("raw_encoded_url_string: "+str(raw_encoded_url_string))
if raw_encoded_url_string.rfind('-') == -1:
	decoded_url_string=urllib.parse.unquote_to_bytes(raw_encoded_url_string)
	#print(str(decoded_url_string))
	if len(decoded_url_string) != 20:
		print('error')
	temp=str(binascii.b2a_hex(decoded_url_string))
	#print(temp)
	decoded_url_string=temp[2:len(temp)-1]
elif raw_encoded_url_string.rfind('-') != -1:
	prefix=raw_encoded_url_string[:raw_encoded_url_string.rfind('-')+1]
	#print(str(len(prefix))+prefix)
	peer_id_part=raw_encoded_url_string[raw_encoded_url_string.rfind('-')+1:]
	decoded_url_string=urllib.parse.unquote_to_bytes(peer_id_part)
	#print(str(decoded_url_string))
	if len(prefix)+len(decoded_url_string) != 20:
		print('error')
	temp=str(binascii.b2a_hex(decoded_url_string))
	temp=temp[2:len(temp)-1]
	#print(temp)
	decoded_url_string=prefix+temp
print("decoded_url_string: "+decoded_url_string)



print()
raw_encoded_url_string='S58B-----%b9b%2cs%c3%15%c6Q%3cu%95%96'

print("raw_encoded_url_string: "+str(raw_encoded_url_string))
if raw_encoded_url_string.rfind('-') == -1:
	decoded_url_string=urllib.parse.unquote_to_bytes(raw_encoded_url_string)
	#print(str(decoded_url_string))
	if len(decoded_url_string) != 20:
		print('error')
	temp=str(binascii.b2a_hex(decoded_url_string))
	#print(temp)
	decoded_url_string=temp[2:len(temp)-1]
elif raw_encoded_url_string.rfind('-') != -1:
	prefix=raw_encoded_url_string[:raw_encoded_url_string.rfind('-')+1]
	#print(str(len(prefix))+prefix)
	peer_id_part=raw_encoded_url_string[raw_encoded_url_string.rfind('-')+1:]
	decoded_url_string=urllib.parse.unquote_to_bytes(peer_id_part)
	#print(str(decoded_url_string))
	if len(prefix)+len(decoded_url_string) != 20:
		print('error')
	temp=str(binascii.b2a_hex(decoded_url_string))
	temp=temp[2:len(temp)-1]
	#print(temp)
	decoded_url_string=prefix+temp
print("decoded_url_string: "+decoded_url_string)



print()
#need to store these in hex
ip4_address='128.168.0.10'
ip_port=8080
#print(ip_port)
ip_port_in_hex=str(hex(ip_port))[2:]
while len(ip_port_in_hex) !=4:
	ip_port_in_hex='0'+ip_port_in_hex
#print(ip_port_hex)
#if dealing with an ipv4 address, then count the octets
#split into 4 sections
#convert to hex, combine the hex pieces+hexed port
if ip4_address.count('.') == 3:
	#dealing with an ipv4 address
	ip_list=ip4_address.split('.')
	for i in (range(len(ip_list))):
		#print(str(ip_list[i]))
		ip_list[i]=hex(int(ip_list[i]))
		if len(ip_list[i]) != 4:
			ip_list[i]=ip_list[i][:2]+'0'+ip_list[i][2:]
		#print(str(ip_list[i]))
		#temp=str(ip_list[0])[2:]
		#temp1=str(ip_list[1])[2:]
		#temp2=str(ip_list[2])[2:]
		#temp3=str(ip_list[3])[2:]
		#address_no_port=temp+temp1+temp2+temp3
	address_in_hex_no_port=str(ip_list[0])[2:]+str(ip_list[1])[2:]+str(ip_list[2])[2:]+str(ip_list[3])[2:]
else:
	#dealing with an ipv6 address
	address_no_port=ip4_address
#print(ip4_address+":"+str(ip_port))
#print(address_no_port)
hex_address=address_in_hex_no_port+ip_port_in_hex
#print(hex_address)

raw_encoded_url_string3='S58B-----%15%c6%3cu%95%96'

#print()
#print("raw_encoded_url_string3: "+str(raw_encoded_url_string3))
if raw_encoded_url_string3.rfind('-') != -1:
	prefix=raw_encoded_url_string3[:raw_encoded_url_string3.rfind('-')+1]
	#print(prefix)
	peer_id_part=raw_encoded_url_string3[raw_encoded_url_string3.rfind('-')+3:]
	decoded_url_string3=binascii.b2a_hex(urllib.parse.unquote_to_bytes(peer_id_part))
	decoded_url_string3=prefix+str(decoded_url_string3)
else:
	decoded_url_string3=binascii.b2a_hex(urllib.parse.unquote_to_bytes(raw_encoded_url_string3))
#print("decoded_url_string3: " + str(len(decoded_url_string3)) + ": " + str(decoded_url_string3))

def createTable(table_id,client_dictionary,client_remote_ip):
        #use the client dictionary along with IP info to create a Peer
        #determine if a client_ip was used in the client_dictionary, if so use it, else use the IP from the connection instead

        if (client_dictionary['left'] == 0):
            isSeed=True
        elif (client_dictionary['left'] != 0):
            isSeed=False
        
        if client_ip in client_dictionary:
            new_Peer=createPeer(client_dictionary['client_peer_id'],isSeed,client_dictionary['client_ip'],client_dictionary['client_port'])
        elif client_ip not in client_dictionary:
            new_Peer=createPeer(client_dictionary['client_peer_id'],isSeed,client_remote_ip ,client_dictionary['client_port'])

        #set initial values for peer count
        complete=0
        incomplete=0
        if (isSeed == True):
            complete=1
        elif (isSeed != True):
            incomplete=1
        
        #and finally create a new table using the table id specified and the new Peer
        #generate new peer
        
        #table_id[0].append()
        return



info_hash_table=({})
complete=0
incomplete=0
client0=['-UT2210-b9b2csc315c6Q3cu9596',False,'192.168.0.1',2342,'192.168.0.1',2342]
client1=['-UT2210-b9b2csc315c6Q3cu9595',False,'192.168.0.2',9653,'192.168.0.1',2342]
client2=['-UT2210-b9b2csc315c6Q3cu9593',False,'192.168.0.3',5412,'192.168.0.1',2342]

#print(client0)
#print(client1)
#print(client2)

info_hash_table['complete']=complete
info_hash_table['incomplete']=incomplete

#map the md5 hash of the peer_id to the client object
clientHash=hashlib.md5()
clientHash.update(client0[0].encode('utf-8'))
info_hash_table[clientHash.digest()]=client0
clientHash=hashlib.md5()
clientHash.update(client1[0].encode('utf-8'))
info_hash_table[clientHash.digest()]=client1
clientHash=hashlib.md5()
clientHash.update(client2[0].encode('utf-8'))
info_hash_table[clientHash.digest()]=client2

#print("Entries in info_hash_table: " +str(len(info_hash_table)))
#print("complete:"+ str(info_hash_table["complete"]))
#print("incomplete:" + str(info_hash_table["incomplete"]))


#query a single client
#print("client_ip:")
client_query=hashlib.md5()
client_query.update(client0[0].encode('utf-8'))
#print(info_hash_table[client_query.digest()][2])
#print("client port")
client_query=hashlib.md5()
client_query.update(client0[0].encode('utf-8'))
#print(info_hash_table[client_query.digest()][3])

#build a peerlist
#md5 hashes of the peer_id are used as keys
info_hashes_table_keys=info_hash_table.keys()
peer_list=[]
for i in (info_hashes_table_keys):
	if (i != 'complete'):
		if (i != 'incomplete'):
			#print(str(info_hash_table[i][2])+":"+str(info_hash_table[i][3]))
			peer_list.append((info_hash_table[i][2],info_hash_table[i][3]))

#print (peer_list)

master_info_hash_table=({})
#masterlist is a dictionary mapping md5 hashes of info_hashes to table_objects (dict objects)
info_hash0='%124Vx%9A%BC%DE%F1%23Eg%89%AB%CD%EF%124Vx%9A '
info_hash1='%124Vx%9A%BC%DE%F1%23Eg%89%AB%CD%EF%124Vx%9B '
info_hash2='%124Vx%9A%BC%DE%F1%23Eg%89%AB%CD%EF%124Vx%9C '

#Class dummy();
#	pass

	
#client_requested_info_hash=hashlib.md5()
#client_requested_info_hash.update(info_hash0.encode('utf-8'))
#info_hash_table[clientHash.digest()]=client0
#if not exist then create a new one
#if client_requested_info_hash.digest() not in master_info_hash_table:
#	master_info_hash_table[client_requested_info_hash.digest()]=createTable()
#setattr(master_info_hash_table[client_requested_info_hash.digest()],newTable,newTable)
	#dummy_reference_table={newTable:master_info_hash_table[client_requested_info_hash.digest()]}
	#master_info_hash_table[client_requested_info_hash.digest()].set_name(newTable)
#elif client_requested_info_hash.digest() in master_info_hash_table:
#	print("already present")



#this should be the table string now
#print(master_info_hash_table[client_requested_info_hash.digest()])
#this should be the table object now
#print(str(master_info_hash_table[client_requested_info_hash.digest()]))

